engine/src/main/java/org/apache/apex/log/appender/ApexRollingFileAppender.java

```diff
@@ -0,0 +1,142 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.log.appender;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+import org.apache.log4j.RollingFileAppender;
+import org.apache.log4j.helpers.CountingQuietWriter;
+import org.apache.log4j.helpers.LogLog;
+
+public class ApexRollingFileAppender extends RollingFileAppender
+{
+  private int currentFileIndex = 0;
+  private String currentFileName = "";
+
+  public synchronized void setFile(String fileName, boolean append, boolean bufferedIO, int bufferSize)
+      throws IOException
+  {
+    currentFileName = fileName + "." + currentFileIndex;
+    LogLog.debug("setFile called: " + currentFileName + ", " + append);
+
+    // It does not make sense to have immediate flush and bufferedIO.
+    if (bufferedIO) {
+      setImmediateFlush(false);
+    }
+
+    reset();
+    FileOutputStream ostream = null;
+    try {
+      //   attempt to create file
+      ostream = new FileOutputStream(currentFileName, append);
+    } catch (FileNotFoundException ex) {
+      //   if parent directory does not exist then
+      //      attempt to create it and try to create file
+      //      see bug 9150
+      String parentName = new File(currentFileName).getParent();
+      if (parentName != null) {
+        File parentDir = new File(parentName);
+        if (!parentDir.exists() && parentDir.mkdirs()) {
+          ostream = new FileOutputStream(currentFileName, append);
+        } else {
+          throw ex;
+        }
+      } else {
+        throw ex;
+      }
+    }
+    Writer fw = createWriter(ostream);
+    if (bufferedIO) {
+      fw = new BufferedWriter(fw, bufferSize);
+    }
+
+    if (new File(fileName).exists()) {
+      Files.delete(Paths.get(fileName));
+    }
+    Files.createSymbolicLink(Paths.get(fileName), Paths.get(currentFileName));
+    this.setQWForFiles(fw);
+    this.fileName = fileName;
+    this.fileAppend = append;
+    this.bufferedIO = bufferedIO;
+    this.bufferSize = bufferSize;
+    writeHeader();
+    LogLog.debug("setFile ended");
+  }
+
+  @Override
+  public void rollOver()
+  {
+    File file;
```

Comment:
Why does this file variable need to be declared here when it's only used within the if (maxBackupIndex > 0)